File: Source/Core/Configurations/MirageConfiguration.vb
Imports System.IO
Imports System.Windows.Forms
Imports Microsoft.Extensions.Configuration

Public Class MirageConfiguration
    Implements IDisposable

    Private isDisposed As Boolean
    Private ReadOnly configuration As IConfigurationRoot

    Public Sub New(Optional envPrefix As String = "MIRAGE")
        Dim currentEnvironment As String = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT")

        Dim files As IEnumerable(Of String) = Directory.GetFiles(AppContext.BaseDirectory) _
            .Where(Function(name) name.Contains("appsettings")) _
            .Where(Function(name) Not name.Contains(".development")) _
            .Where(Function(name) Not name.Contains(".production")) _
            .Where(Function(name) name.EndsWith(".json"))

        Dim envfiles As IEnumerable(Of String) = Directory.GetFiles(AppContext.BaseDirectory) _
            .Where(Function(name) name.Contains("appsettings")) _
            .Where(Function(name) name.EndsWith($".{currentEnvironment}.json"))

        Dim builder As IConfigurationBuilder = New ConfigurationBuilder()

        For Each file As String In files
            Console.WriteLine($"Reading configuration file '{file}'...")
            builder = builder.AddJsonFile(file, optional:=True, reloadOnChange:=True)
        Next

        For Each file As String In envfiles
            Console.WriteLine($"Reading configuration file '{file}'...")
            builder = builder.AddJsonFile(file, optional:=True, reloadOnChange:=True)
        Next

        Console.WriteLine($"Reading configuration environment variables with prefix '{envPrefix}'...")
        builder = builder.AddEnvironmentVariables(envPrefix)

        Me.configuration = builder.Build()
    End Sub


    Protected Overrides Sub Finalize()
        Me.Dispose(disposing:=False)
        MyBase.Finalize()
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
        Me.Dispose(disposing:=True)
        GC.SuppressFinalize(Me)
    End Sub

    Protected Overridable Sub Dispose(disposing As Boolean)
        If Not Me.isDisposed Then
            If disposing Then

            End If

            Me.isDisposed = 1
        End If
    End Sub

    Public Function GetValue(Of ValueType)(key As String, defaultValue As ValueType) As ValueType
        Dim value As String = Me.configuration(key)

        If Not String.IsNullOrWhiteSpace(value) Then
            Try
                Return CType(Convert.ChangeType(value, GetType(ValueType)), ValueType)
            Catch ex As InvalidCastException
                Console.WriteLine($"[Error] Unable to read configuration value '{key}' as type of '{NameOf(ValueType)}'.")
                Return defaultValue
            End Try
        Else
            Console.WriteLine($"[Error] Unable to read configuration value '{key}' as it does not exist'.")
            Return defaultValue
        End If
    End Function

End Class

File: Source/Core/Configurations/MirageConfiguration.cs
using System;
using System.IO;
using System.Linq;
using Microsoft.Extensions.Configuration;
using Microsoft.VisualBasic.CompilerServices;

namespace Core
{

    public class MirageConfiguration : IDisposable
    {

        private bool isDisposed;
        private readonly IConfigurationRoot configuration;

        public MirageConfiguration(string envPrefix = "MIRAGE")
        {
            string currentEnvironment = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");

            var files = Directory.GetFiles(AppContext.BaseDirectory).Where(name => name.Contains("appsettings")).Where(name => !name.Contains(".development")).Where(name => !name.Contains(".production")).Where(name => name.EndsWith(".json"));




            var envfiles = Directory.GetFiles(AppContext.BaseDirectory).Where(name => name.Contains("appsettings")).Where(name => name.EndsWith($".{currentEnvironment}.json"));


            IConfigurationBuilder builder = new ConfigurationBuilder();

            foreach (string @file in files)
            {
                Console.WriteLine($"Reading configuration file '{@file}'...");
                builder = builder.AddJsonFile(@file, optional: true, reloadOnChange: true);
            }

            foreach (string @file in envfiles)
            {
                Console.WriteLine($"Reading configuration file '{@file}'...");
                builder = builder.AddJsonFile(@file, optional: true, reloadOnChange: true);
            }

            Console.WriteLine($"Reading configuration environment variables with prefix '{envPrefix}'...");
            builder = builder.AddEnvironmentVariables(envPrefix);

            configuration = builder.Build();
        }


        ~MirageConfiguration()
        {
            Dispose(disposing: false);
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!isDisposed)
            {
                if (disposing)
                {

                }

                isDisposed = Conversions.ToBoolean(1);
            }
        }

        public ValueType GetValue<ValueType>(string key, ValueType defaultValue)
        {
            string value = configuration[key];

            if (!string.IsNullOrWhiteSpace(value))
            {
                try
                {
                    return Conversions.ToGenericParameter<ValueType>(Convert.ChangeType(value, typeof(ValueType)));
                }
                catch (InvalidCastException ex)
                {
                    Console.WriteLine($"[Error] Unable to read configuration value '{key}' as type of '{nameof(ValueType)}'.");
                    return defaultValue;
                }
            }
            else
            {
                Console.WriteLine($"[Error] Unable to read configuration value '{key}' as it does not exist'.");
                return defaultValue;
            }
        }

    }
}